crates/ruff_linter/src/rules/ruff/rules/sequence_sorting.rs

```diff
@@ -0,0 +1,878 @@
+/// classes), followed by anything else. Within each category,
+/// a [natural sort](https://en.wikipedia.org/wiki/Natural_sort_order)
+/// is used to order the elements.
+struct IsortSortKey<'a> {
+    category: InferredMemberType,
+    value: &'a str,
+}
+
+impl Ord for IsortSortKey<'_> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.category
+            .cmp(&other.category)
+            .then_with(|| natord::compare(self.value, other.value))
+    }
+}
+
+impl PartialOrd for IsortSortKey<'_> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for IsortSortKey<'_> {
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(other) == Ordering::Equal
+    }
+}
+
+impl Eq for IsortSortKey<'_> {}
+
+impl<'a> From<&'a str> for IsortSortKey<'a> {
+    fn from(value: &'a str) -> Self {
+        Self {
+            category: InferredMemberType::of(value),
+            value,
+        }
+    }
+}
+
+impl<'a> From<&'a StringSequenceItem> for IsortSortKey<'a> {
+    fn from(item: &'a StringSequenceItem) -> Self {
+        Self::from(item.value.as_str())
+    }
+}
+
+/// Classification for the casing of an element in a
+/// sequence of literal strings.
+///
+/// This is necessary to achieve an "isort-style" sort,
+/// where elements are sorted first by category,
+/// then, within categories, are sorted according
+/// to a natural sort.
+///
+/// You'll notice that a very similar enum exists
+/// in ruff's reimplementation of isort.
+#[derive(Debug, Ord, PartialOrd, Eq, PartialEq, Clone, Copy)]
+enum InferredMemberType {
+    Constant,
+    Class,
+    Other,
+}
+
+impl InferredMemberType {
+    fn of(value: &str) -> Self {
+        // E.g. `CONSTANT`
+        if value.len() > 1 && is_cased_uppercase(value) {
+            Self::Constant
+        // E.g. `Class`
+        } else if value.starts_with(char::is_uppercase) {
+            Self::Class
+        // E.g. `some_variable` or `some_function`
+        } else {
+            Self::Other
+        }
+    }
+}
+
+/// An enumeration of the various kinds of sequences for which Python has
+/// [display literals](https://docs.python.org/3/reference/expressions.html#displays-for-lists-sets-and-dictionaries).
+///
+/// (I'm aware a set isn't actually a "sequence",
+/// *but* for our purposes it's conceptually a sequence,
+/// since in terms of the AST structure it's almost identical
+/// to tuples/lists.)
+///
+/// Whereas lists, dicts and sets are always parenthesized
+/// (e.g. lists always start with `[` and end with `]`),
+/// single-line tuples *can* be unparenthesized.
+/// We keep the original AST node around for the
+/// Tuple variant so that this can be queried later.
+#[derive(Debug)]
+pub(super) enum SequenceKind<'a> {
+    List,
+    Set,
+    Tuple(&'a ast::ExprTuple),
+}
+
+impl SequenceKind<'_> {
+    fn surrounding_parens(&self, source: &str) -> (&'static str, &'static str) {
```

Comment:
I think we typically call these "brackets", i.e., brackets are the set that includes parentheses (`(`), square brackets (`[`), and braces (`{`).

Reply:
Actually, we don't :wink: . At least not in the formater. We call these `parentheses` (I avoid short forms wherever possible).